Lib/pty.py

```diff
@@ -170,3 +175,128 @@ def spawn(argv, master_read=_read, stdin_read=_read):
 
     os.close(master_fd)
     return os.waitpid(pid, 0)[1]
+
+# Author: Soumendra Ganguly.
+SIGWINCH = signal.SIGWINCH
+
+def _login_pty(master_fd, slave_fd):
+    """Given a pty, makes the calling process a session leader, makes the pty slave
+    its controlling terminal, stdin, stdout, and stderr. Closes both pty ends."""
+    # Establish a new session.
+    os.setsid()
+    os.close(master_fd)
+
+    try:
+        fcntl.ioctl(slave_fd, termios.TIOCSCTTY) # Make the pty slave the controlling terminal.
+    except:
+        os.close(slave_fd)
+        raise
+
+    # Slave becomes stdin/stdout/stderr.
+    os.dup2(slave_fd, STDIN_FILENO)
+    os.dup2(slave_fd, STDOUT_FILENO)
+    os.dup2(slave_fd, STDERR_FILENO)
+    if (slave_fd > STDERR_FILENO):
+        os.close(slave_fd)
+
+def _winresz(pty_slave):
+    """Resize window."""
+    w = struct.pack('HHHH', 0, 0, 0, 0)
+    s = fcntl.ioctl(STDIN_FILENO, termios.TIOCGWINSZ, w)
+    fcntl.ioctl(pty_slave, termios.TIOCSWINSZ, s)
+
+def _create_hwinch(pty_slave):
+    """Creates SIGWINCH handler."""
+    def _hwinch(signum, frame):
+        try:
+            _winresz(pty_slave)
+        except:
+            pass
+    return _hwinch
+
```

Comment:
C code equivalent to "_cleanup" would use close(2), tcsetattr(3), and sigaction(2).

References:
1. https://man7.org/linux/man-pages/man2/close.2.html
2. https://man7.org/linux/man-pages/man3/termios.3.html
3. https://man7.org/linux/man-pages/man2/sigaction.2.html